src/search/SearchMessagesScreen.js

```diff
@@ -131,6 +149,35 @@ class SearchMessagesScreenInner extends PureComponent<Props, State> {
     this.handleQuerySubmit(e);
   };
 
+  fetchOlder = async () => {
+    if (!this.props.loading && !this.state.foundOldest && !this.state.isFetching) {
+      this.setState({ isFetching: true });
+      try {
+        // Not sure why we require a double await here.
+        const { messages: fetchedMessages, foundOldest } = await await this.props.dispatch(
+          fetchMessages({
+            anchor: this.state.messages ? this.state.messages[0].id : LAST_MESSAGE_ANCHOR,
+            narrow: SEARCH_NARROW(this.state.query),
+            numBefore: config.messagesPerRequest,
+            numAfter: 0,
+          }),
+        );
+        // Last fetched message === first current message, hence remove it
+        fetchedMessages.pop();
+        this.setState(prevState => ({
+          messages: [...fetchedMessages, ...prevState.messages],
+          foundOldest,
+        }));
+      } finally {
+        this.setState({ isFetching: false });
```

Comment:
This uses the same state property that's managed by `handleQuerySubmit` to describe whether there's an ongoing fetch for a new search query, and I think it doesn't interact correctly with that -- in particular it can clear `isFetching` when in fact there's still a fetch ongoing for a new search query.

In general it's probably quite tricky to have this managed by both this code and also `handleQuerySubmit`. It's easier to reason about when there's just one pair of lines setting and clearing it.

Fortunately I think there isn't actually a need to have this operate on the same bit of state -- the one existing place that consumes `isFetching`, in `SearchMessagesCard`, only cares about it if there are no messages in the current state, so it should never even notice whether `fetchOlder` has set it. We do want this `fetchOlder` to have a flag that allows it to avoid running twice concurrently, but we can just make it a separate thing.

Reply:
added a non react state `isFetchingOlder` to address this.